pkg/net/libp2p/libp2ptransport.go

```diff
@@ -173,37 +176,44 @@ func (t *Transport) connectToNode(ctx context.Context, addr multiaddr.Multiaddr)
 
 	s, err := t.openStream(ctx, info.ID)
 	if err != nil {
-		t.logger.Log(logging.LevelError, fmt.Sprintf("couldn't open stream: %v", err))
-		return nil, fmt.Errorf("couldn't open stream to %v: %w", addr, err)
+		return nil, fmt.Errorf("failed to open new stream to node %v: %w", addr, err)
 	}
 
 	return s, nil
 }
 
 func (t *Transport) openStream(ctx context.Context, p peer.ID) (network.Stream, error) {
-	for {
-		sctx, cancel := context.WithTimeout(ctx, defaultMaxTimeout)
-		s, err := t.host.NewStream(sctx, p, ID)
+	var streamErr error
+	for i := 0; i < retryAttempts; i++ {
+		sctx, cancel := context.WithTimeout(ctx, maxConnectingTimeout)
+
+		s, streamErr := t.host.NewStream(sctx, p, ProtocolID)
 		cancel()
 
-		if err == nil {
+		if streamErr == nil {
 			return s, nil
 		}
 
-		t.logger.Log(logging.LevelError, fmt.Sprintf("failed to open stream: %v", err))
-
-		delay := time.NewTimer(defaultMaxTimeout)
+		if i >= nonErrorAttempts {
+			t.logger.Log(
+				logging.LevelError, fmt.Sprintf("failed to open stream to %s, retry in %d sec", p, retryTimeout))
+		} else {
+			t.logger.Log(
+				logging.LevelInfo, fmt.Sprintf("failed to open stream to %s, retry in %d sec", p, retryTimeout))
+		}
 
+		delay := time.NewTimer(retryTimeout)
 		select {
 		case <-delay.C:
 			continue
 		case <-ctx.Done():
 			if !delay.Stop() {
```

Comment:
Similar to previous one.
It seems we don't need to stop the timer here, since we're returning anyway, so it would seem much cleaner to me to ignore the timer here.

Reply:
This is a best practice to prevent memory leak - https://www.arangodb.com/2020/09/a-story-of-a-memory-leak-in-go-how-to-properly-use-time-after/

Reply:
Fair enough. I still wouldn't really call it a memory leak in this case, since the memory always reclaimed after the timer fires, it just might not happen immediately. But if this is best practice, let's keep it.